grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/GrpcWebTrailersExtractor.java

```diff
@@ -141,40 +113,93 @@ protected HttpObject filter(HttpObject obj) {
                                 DecompressorRegistry.getDefaultInstance().lookupDecompressor(grpcEncoding);
                         if (decompressor == null) {
                             // Can't find decompressor.
-                            deframer.close();
+                            publisher.close();
                             return obj;
                         }
-                        deframer.decompressor(ForwardingDecompressor.forGrpc(decompressor));
+                        handler.decompressor(ForwardingDecompressor.forGrpc(decompressor));
                     }
-                    deframer.request(Integer.MAX_VALUE);
                     return obj;
                 }
 
-                if (obj instanceof HttpData && !deframer.isClosed()) {
+                if (obj instanceof HttpData && !publisher.isComplete()) {
                     final HttpData httpData = (HttpData) obj;
                     final HttpData wrapped = HttpData.wrap(
                             httpData.byteBuf(ByteBufAccessMode.RETAINED_DUPLICATE));
-                    deframer.deframe(wrapped, wrapped.isEndOfStream());
+                    final boolean ignored = publisher.tryWrite(wrapped);
                 }
                 return obj;
             }
 
             @Override
             protected void beforeComplete(Subscriber<? super HttpObject> subscriber) {
-                deframer.close();
+                publisher.close();
             }
 
             @Override
             protected Throwable beforeError(Subscriber<? super HttpObject> subscriber, Throwable cause) {
-                deframer.close();
+                publisher.close();
                 return cause;
             }
         };
         filteredHttpResponse.whenComplete().handle((unused, unused2) -> {
             // To make sure the deframer is closed even when the response is cancelled.
-            deframer.close();
+            publisher.close();
             return null;
         });
         return filteredHttpResponse;
     }
+
+    private static Subscriber<DeframedMessage> trailersSubscriber(ClientRequestContext ctx) {
+        return new Subscriber<DeframedMessage>() {
```

Comment:
Could be a singleton?

Reply:
Unfortunately, not. This anonymous `Subscriber` needs to access `ClientRequestContext`.
Let me introduce `TrailersSubscriber` class for readability.

Reply:
Ah, OK.